src/charm.py

```diff
@@ -302,20 +369,212 @@ def _on_admin_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
             logger.info("This app's admin ingress URL: %s", event.url)
 
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
 
     def _on_public_ingress_ready(self, event: IngressPerAppReadyEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app's public ingress URL: %s", event.url)
 
         self._handle_status_update_config(event)
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
 
     def _on_ingress_revoked(self, event: IngressPerAppRevokedEvent) -> None:
         if self.unit.is_leader():
             logger.info("This app no longer has ingress")
 
         self._handle_status_update_config(event)
         self._update_hydra_endpoints_relation_data(event)
+        self._update_endpoint_info()
+
+    def _on_oauth_relation_created(self, event: RelationCreatedEvent) -> None:
+        self._update_endpoint_info()
+
+    def _on_client_created(self, event: ClientCreatedEvent) -> None:
```

Comment:
There seems to be missing validations that the unit is the leader. For instance we probably don't want all units to create clients and store the mapping in the peer relation data.

Reply:
There also seems to be lacking validation that the peer relation was created before adding/removing data from it.

Reply:
I looked at other charms code to do this and they weren't checking if the peer relation was ready.
How would you do that? Is something like this enough?
```python
if not self.model.get_relation(PEER):
    return
```

Reply:
I thing `get_relation` throws an error when the relation name doesn't exist but yes somehting along those lines